examples/Tests/EQDiagram.hpp

```diff
@@ -0,0 +1,267 @@
+    }
+  }
+};
+
+/* 
+    UI Object to display values computed by an EQDiagramCalculator. When initialized with halp::custom_item_base<EQDiagram>, xInput, yInput and valuesNb should be specified
+    to match an EQDiagramCalculator's frequencies, responses and plotAmount value.
+*/
+template <int w, int h, int N>
+struct EQDiagram
+{
+  EQDiagram()
+      : valuesNb(N)
+  {
+    //maybe should fill with coords to draw flatline at start
+    abscissa.resize(N);
+    ordinate.resize(N);
+  }
+  static constexpr double width() { return w; }
+  static constexpr double height() { return h; }
+  static constexpr double layout()
+  {
+    enum
+    {
+      custom
+    } d{};
+    return d;
+  }
+
+  // double x = 0.0;
+  // double y = 0.0;
+
+  int valuesNb;
+
+  // the followings will hold the coords of plotted points
+  std::vector<float> abscissa{};
+  std::vector<float> ordinate{};
+
+  std::vector<float> xticks{20, 500, 1, 2, 5, 10, 20};
+  std::vector<std::pair<float, float>> yticks;
+
+  // basically construct a log scale for x values and compute coords to correctly plot values in the UI
+
+  std::function<float(float)> xScale = [](float val)
+  {
+    auto xMin = log10(20);
+    auto xMax = log10(20000);
+    auto xscale = width() / (xMax - xMin);
+    auto xdiff = -xscale * xMin;
+    return xscale * log10(val) + xdiff;
+  };
+
+  void update(const float* xInput, const float* yInput)
+  {
+    //updating plot values
+    auto yMin = utilities::min(yInput, valuesNb);
+    auto yMax = utilities::max(yInput, valuesNb);
+    auto yscale = height() / utilities::absolute(yMax - yMin);
+    auto ydiff = -yscale * yMax;
+    for (auto i = 0; i < valuesNb; ++i)
+    {
+      abscissa[i] = xScale(*(xInput + i));
+      ordinate[i] = -(yscale * (*(yInput + i)) + ydiff);
+    }
+
+    // updating y axis scale labels
+    auto ticksNb = 10;
+    auto step = (yMax - yMin) / (ticksNb - 1);
+    for (auto i = 0; i < ticksNb; ++i)
+    {
+      yticks.push_back({yMin + i * step, -(yscale * (yMin + i * step) + ydiff)});
+    }
+  }
+
+  void paint(avnd::painter auto ctx)
+  {
+    // THE AXES
+    // x axis
+    ctx.set_stroke_color({.r = 103, .g = 103, .b = 103, .a = 255});
+    ctx.set_fill_color({255, 255, 255, 255});
+    ctx.begin_path();
+    ctx.draw_line(0, height() - 1, width(), height() - 1);
+    ctx.draw_line(0, height(), width(), height());
+    ctx.fill();
+    ctx.stroke();
+    for (auto i = 0; i < 2; ++i)
+    {
+      // scale labels
+      auto atX = xScale(xticks[i]);
+      ctx.begin_path();
+      ctx.draw_text(atX, height() - 10., utilities::to_string(xticks[i], 0));
+      ctx.fill();
+
+      // vertical lines
+      ctx.begin_path();
+      ctx.draw_line(atX, 0, atX, height());
+      ctx.fill();
+      ctx.stroke();
+    }
+    for (auto i = 2; i < xticks.size(); ++i)
+    {
+      // scale labels
+      auto atX = xScale(xticks[i] * 1000);
+      ctx.begin_path();
+      std::string grad = utilities::to_string(xticks[i], 0) + "k";
```

Comment:
directly use fmt::format: 

    ctx.draw_text(..., fmt::format("{.0f}k", xticks[i]));

or even better, store them somewhere to reuse them as string formatting is always an expensive operation - and an open research problem for floating-point

Reply:
since the xticks are fixed, I eventually made an array of ```std::pair<float, std::string_view>``` so the strings are now reusable (and code is bit cleaner :))

Reply:
good job !